packages/terra-pills/src/Pill.jsx

```diff
@@ -0,0 +1,212 @@
+   * Only applies when used along with the 'onSelect' prop.
+   */
+  ariaControls: PropTypes.string,
+  /**
+   * Tooltip to display if the pill label does not have enough space to display and will show as truncated, to be used
+   * in addition to a popup. Only applies when used along with the 'onSelect' prop.
+   */
+  title: PropTypes.string,
+  /**
+   * @private
+   * The intl object to be injected for translations.
+   */
+  intl: PropTypes.shape({ formatMessage: PropTypes.func }).isRequired,
+};
+
+const defaultProps = {
+  id: undefined,
+  onRemove: undefined,
+  onSelect: undefined,
+  refCallback: undefined,
+  ariaHasPopup: false,
+  ariaExpanded: false,
+  ariaControls: undefined,
+  title: undefined,
+};
+
+const Pill = (props) => {
+  const {
+    label,
+    id,
+    onRemove,
+    onSelect,
+    refCallback,
+    ariaHasPopup,
+    ariaExpanded,
+    ariaControls,
+    title,
+    intl,
+    ...customProps
+  } = props;
+
+  const pillRef = useRef();
+
+  useEffect(() => {
+    if (refCallback) {
+      refCallback(pillRef.current);
+    }
+  }, [refCallback]);
+
+  const handleOnClick = () => {
+    onSelect();
+  };
+
+  const handleOnRemove = () => {
+    onRemove();
+  };
+
+  const handleOnKeyDown = (event) => {
+    pillRef.current.setAttribute('data-terra-pills-show-focus-styles', 'true');
+    if ((event.keyCode === KEY_RETURN || event.keyCode === KEY_SPACE) && onSelect) {
+      onSelect();
+      event.preventDefault();
+    } else if ((event.keyCode === KEY_DELETE || event.keyCode === KEY_BACK_SPACE) && onRemove) {
+      onRemove();
+    }
+  };
+
+  const handleOnMouseDown = () => {
+    pillRef.current.setAttribute('data-terra-pills-show-focus-styles', 'false');
+  };
+
+  const pillProps = {};
+  if (onSelect || onRemove) {
+    pillProps.tabIndex = '0';
+    pillProps.onKeyDown = handleOnKeyDown;
+    pillProps.onMouseDown = handleOnMouseDown;
+    pillProps.role = 'button';
+  }
+
+  const pillButtonProps = {};
+  if (onSelect) {
+    pillButtonProps.title = title;
+    pillButtonProps.onClick = handleOnClick;
+  }
+
+  const removeButtonProps = {};
+  if (onRemove) {
+    removeButtonProps.onClick = handleOnRemove;
+  }
+
+  let pillInteractionHint;
+  if (onSelect && onRemove) {
+    pillInteractionHint = intl.formatMessage({ id: 'Terra.pills.pillHint.selectableAndRemovable' });
+  } else if (onSelect) {
+    pillInteractionHint = intl.formatMessage({ id: 'Terra.pills.pillHint.selectable' });
+  } else if (onRemove) {
+    pillInteractionHint = intl.formatMessage({ id: 'Terra.pills.pillHint.removable' });
+  }
+
+  const theme = React.useContext(ThemeContext);
+
+  const pillClassNames = classNames(
+    cx([
+      'pill-container',
+      { 'is-focusable': !!onSelect || !!onRemove },
+      { 'is-selectable': !!onSelect },
+      { 'is-selectable-and-removable': !!onSelect && !!onRemove },
```

Comment:
nit : can we move these common selectable and removable classes to new variable and re-use them on both `pillClassNames` and `pillLabelClassNames` 
```js
  const selectableAndRemovableClassNames = cx([
    { 'is-selectable': !!onSelect },
    { 'is-removable': !!onRemove },
    { 'is-selectable-and-removable': !!onSelect && !!onRemove },
  ]);

  const pillClassNames = classNames(
    cx([
      'pill-container',
       selectableAndRemovableClassNames,
      theme.className,
    ]),
    customProps.className,
  );

  const pillLabelClassNames = cx([
    'pill-label',
     selectableAndRemovableClassNames,
  ]);
```